app/src/main/res/values/strings.xml

```diff
@@ -15,6 +15,8 @@
     <string name="permission_location">Will you allow %1$s to access your location?</string>
     <!-- Parameter will be replaced with the app name -->
     <string name="permission_notification">Will you allow %1$s to send notifications?</string>
+    <!-- Parameter will be replaced with the app name -->
+    <string name="permission_read_external_storage">Will you allow %1$s to read your external storage?</string>
```

Comment:
I guess I still find the wording confusing because the FxR dialog isn't actually granting permission, it is only asking the user if they want FxR to ask for permission. Then a system dialog is show (at least on Oculus) that requests that the user grant permission.

Reply:
what is the benefit/rationale to introducing the initial permission prompt?

ideally, we should follow the [Permissions API standard](https://w3c.github.io/permissions/#examples), correct?

Reply:
The [Android recommended flow](https://developer.android.com/training/permissions/requesting#explain) for permissions request includes showing a dialog so you have a chance to explain to the users why you need that permissions before they are prompted by the system.

In our case there are two different sets of permissions, the ones requested by the website and the ones requested by the app. 
Site permissions are requested per site. Gecko will grant permission to each site based on our custom dialog response. The system dialog is only shown once if the permission is already granted.
App permissions are only shown once for the whole app so for example if the user wants to access the internal storage to load a page it will prompted only once in case the user grants the permission.
In both cases we show the same dialog (that requires redesign btw), we just change the requester:
"Will you allow X to Y?" where X can be either "Firefox Reality" or a Website URL.

Oculus Browser does the same thing as us. It shows a banner at the bottom of the page saying that "xyz.com want's to use your location" if you say "Ok" the system dialog is shown.

Samsung Internet request permissions upon first start and doesn't let you use the browser unless you grant both external storage and location (for all sites) which is not the recommended flow as you should only request permissions when the are needed.

Reply:
thanks for writing this all up. this is all helpful information.

ah, I understand now. and, you're right, Samsung Internet does prompt upon a fresh install -> launch for the File read permissions.

![image](https://user-images.githubusercontent.com/203725/46770480-6a9a4c80-cca4-11e8-9c71-53c2e807aeb5.png)

![image](https://user-images.githubusercontent.com/203725/46770484-6cfca680-cca4-11e8-92be-9f77a6046350.png)

see my [review comments](https://github.com/MozillaReality/FirefoxReality/pull/605#pullrequestreview-163591570), though, because I have the accepted the File read permissions.

another question: does Android conflate the "Storage" permission to mean both read _and_ **write** access? because when I granted the Android permissions, I noticed that both permissions got granted (from running `adb shell dumpsys package org.mozilla.vrbrowser`):

    ```
          runtime permissions:
            android.permission.READ_EXTERNAL_STORAGE: granted=true
            android.permission.WRITE_EXTERNAL_STORAGE: granted=true
    ```

I expected it to set only `android.permission.READ_EXTERNAL_STORAGE`.

Reply:
Don't have an explanation for this but for some reason it seems to be Oculus related. It is not happening in the noapi or Daydream flavors. I've checked the merged manifests and I can't find a reason so it might be something that Oculus is doing under the hood.